cluster/images/etcd-version-monitor/etcd-version-monitor.go

```diff
@@ -0,0 +1,275 @@
+	ClusterVersion		string `json:"etcdcluster"`
+}
+
+// Function for periodically fetching etcd version info.
+func getVersion() {
+	// Create the get request for the etcd endpoint.
+	req, err := http.NewRequest("GET", etcdVersionScrapeURI, nil)
+	if err != nil {
+		log.Fatal("NewRequest: ", err)
+		return
+	}
+
+	for {
+		// Send the get request and receive a response.
+		client := &http.Client{}
+		resp, err := client.Do(req)
+		if err != nil {
+			log.Fatal("Do: ", err)
+			time.Sleep(scrapeTimeout)
+			continue
+		}
+
+		// Obtain EtcdVersion from the JSON response.
+		var version EtcdVersion
+		if err := json.NewDecoder(resp.Body).Decode(&version); err != nil {
+			log.Fatal("JSON decoder for etcd version: ", err)
+			time.Sleep(scrapeTimeout)
+			continue
+		}
+		resp.Body.Close()
+
+		// Record the info in the prometheus metric through labels.
+		etcdVersionFetchCount.With(prometheus.Labels{
+			"server_version": version.ServerVersion,
+			"cluster_version": version.ClusterVersion,
+		}).Inc()
+		time.Sleep(scrapeTimeout)
+	}
+}
+
+// Function for periodically fetching etcd versioned data size from etcdctl.
+func getVersionedDataSize() {
+	for {
+		// Prepare the etcdctl command to be executed.
+		cmd := exec.Command("/bin/etcdctl", "--endpoints=localhost:2379", "endpoint", "status")
+		env := os.Environ()
+		env = append(env, fmt.Sprintf("ETCDCTL_API=3"))
+		cmd.Env = env
+		cmdOut, _ := cmd.StdoutPipe()
+
+		// Run the etcdctl command and get the output.
+		if err := cmd.Start(); err != nil {
+			log.Fatal("Failed to start etcdctl command: ", err)
+			time.Sleep(scrapeTimeout)
+			continue
+		}
+		out, _ := ioutil.ReadAll(cmdOut)
+		if err := cmd.Wait(); err != nil {
+			log.Fatal("Failed to finish etcdctl command: ", err)
+			time.Sleep(scrapeTimeout)
+			continue
+		}
+
+		// Obtain the versioned data size from the comma-seperated output.
+		params := strings.Split(string(out), ",")
+		version := strings.TrimSpace(params[2])
+		size_params := strings.Split(strings.TrimSpace(params[3]), " ")
+		value, _ := strconv.ParseFloat(size_params[0], 64)
+		unit := size_params[1]
+
+		// Record the versioned data size in the prometheus metric with labels.
+		etcdVersionedDataSize.With(prometheus.Labels{
+			"backend_version": "3",
+			"unit": unit,
+			"etcdctl_version": version,
+		}).Set(value)
+		time.Sleep(scrapeTimeout)
+	}
+}
+
+// Struct for unmarshalling the json for etcd grpc request labels.
+type GRPCLabels struct {
+	GrpcMethod			string `json:"grpc_method"`
+	GrpcService			string `json:"grpc_service"`
+}
+
+// Struct for unmarshalling the json for etcd grpc requests metric.
+type EtcdGRPCRequestCount struct {
+	GrpcLabels			GRPCLabels 	`json:"labels"`
+	Count				string 		`json:"value"`
+}
+
+// Function for periodically fetching etcd GRPC request counts.
+func getGRPCRequestCount() {
+	// Map for storing last obtained request count for a given grpc request type.
+	lastCount := make(map[GRPCLabels]float64)
+
+	for {
```

Comment:
The same comment as above - make this function only handle a single request, and another function that will be calling this one in a loop (and handling errors potentially)